[SPARK-42953][Connect] Typed filter, map, flatMap, mapPartitions

zhenlineo opened the pull request:
### What changes were proposed in this pull request?
Implemented new missing methods in the client Dataset API: filter, map, flatMap, mapPartitions.

### Why are the changes needed?
Missing APIs

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Integration tests.
The UDF test does not work with maven.

zhenlineo committed:
Impl typed map etc

zhenlineo commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/UserDefinedFunctionE2ETestSuite.scala:
```diff
@@ -0,0 +1,120 @@
+ * these tests only works with SBT for now.
+ */
+class UserDefinedFunctionE2ETestSuite extends RemoteSparkSession {
+  test("simple udf") {
```
@vicennial I need your help to fix this for maven.

LuciferYang commented:
```
TODO: Need to fix the test for maven.
```

Should we make this one 3.5.0 only? I think Maven test failure will become a blocker for 3.4.0 release

LuciferYang commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala:
```diff
@@ -43,7 +45,27 @@ object IntegrationTestUtils {
+        // Redirect server log into console
+        "--conf",
+        s"spark.driver.extraJavaOptions=-Dlog4j.configuration=$log4j2",
+        "--conf",
```
should be `log4j.configurationFile` for log4j2

zhenlineo replied:
I tried `s"spark.driver.extraJavaOptions=-Dlog4j.configurationFile=$log4j2"`, but the logs is not picked up? Do you know what is the correct way to use `log4j.configurationFile`?

LuciferYang commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala:
```diff
@@ -43,7 +45,27 @@ object IntegrationTestUtils {
+        "--conf",
+        s"spark.driver.extraJavaOptions=-Dlog4j.configuration=$log4j2",
+        "--conf",
+        s"spark.executor.extraJavaOptions=-Dlog4j.configuration=$log4j2")
```
`SimpleSparkConnectService` start with `local` module, really need to configure `spark.executor.extraJavaOptions`?

LuciferYang commented on connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -2468,6 +2470,113 @@ class Dataset[T] private[sql] (
+   * `true`.
+   *
+   * @group typedrel
+   * @since 3.4.0
```
I suggest changing to `@since 3.5.0`

xinrong-meng commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:
```diff
@@ -482,27 +482,63 @@ class SparkConnectPlanner(val session: SparkSession) {
+    commonUdf.getFunctionCase match {
+      case proto.CommonInlineUserDefinedFunction.FunctionCase.SCALAR_SCALA_UDF =>
+        transformTypedMapPartitions(commonUdf, baseRel)
       case _ =>
```
nit: I am wondering if we shall use an explicit
```
case proto.CommonInlineUserDefinedFunction.FunctionCase.PYTHON_UDF
```
and add a `case _` to block other `function` field values.

LuciferYang replied:
+1

LuciferYang commented on connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/UdfUtils.scala:
```diff
@@ -0,0 +1,46 @@
+ * cause Dataset sync conflicts when used together with UDFs. Thus we define them outside, in the
+ * client package.
+ */
+object UdfUtils {
```
A public API or Internal utils?

zhenlineo committed:
address feedbacks

amaliujia commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/IntegrationTestUtils.scala:
```diff
@@ -43,7 +45,25 @@ object IntegrationTestUtils {
-  private[connect] val isDebug = System.getProperty(DEBUG_SC_JVM_CLIENT, "false").toBoolean
+
+  private[connect] lazy val debugConfig: Seq[String] = {
+    val log4j2 = s"$sparkHome/connector/connect/client/jvm/src/test/resources/log4j2.properties"
```
where is `$sparkHome` defined?

zhenlineo replied:
It is at line 42 in this file. We obtained it from env var SPARK_HOME.

amaliujia commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:
```diff
@@ -482,27 +482,66 @@ class SparkConnectPlanner(val session: SparkSession) {
+  private def generateObjAttr[T](enc: ExpressionEncoder[T]): Attribute = {
+    val dataType = enc.deserializer.dataType
+    val nullable = !enc.clsTag.runtimeClass.isPrimitive
+    AttributeReference("obj", dataType, nullable)()
```
How is `obj` defined as an attribute? This looks like a magic name?

zhenlineo replied:
This code logic is in objects.scala. I basically expanded the logic there to be able to accept the encoders directly. The original code used "obj". This is fine as it is only used in the middle of ser/deser to provide the datatype and nullability info.

amaliujia replied:
To confirm this won't cause duplicate attribute issue if there is another attribute come from child plan?

zhenlineo committed:
Fix filter python test errors.

hvanhovell commented on connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -2468,6 +2470,113 @@ class Dataset[T] private[sql] (
+    sparkSession.newDataset[T](encoder) { builder =>
+      builder.getFilterBuilder
+        .setInput(plan.getRoot)
+        .setCondition(udf.apply().expr)
```
This makes me a tad nervous. I think it is possible for users to provide a udf without arguments.  Can we use some marker expression here? UnresolvedStar would be fine I guess.

zhenlineo committed:
Always send the udf input and return types as they are.

For filtered function. If the input arguements are absent but the input udf types are set, then it means it is a typed filter function to filter on all inputs.

hvanhovell commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/UserDefinedFunctionE2ETestSuite.scala:
```diff
@@ -0,0 +1,120 @@
+class UserDefinedFunctionE2ETestSuite extends RemoteSparkSession {
+  test("simple udf") {
+    def dummyUdf(x: Int): Int = x + 5
+    val myUdf = udf(dummyUdf _)
```
For the record this requires us to capture the context because we are using a function in the enclosing class. We definitely should test this, however the more canonical way of writing this would be:
```scala
val myUdf = udf { x: int = >
  x + 5
}
```

zhenlineo replied:
I moved the original udf test from E2E into this class, as all tests in this class cannot run with maven. So I would leave this to @vicennial to address this comment.

hvanhovell reviewed (changes requested):
Looks good. Some small comments.

zhenlineo commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:
```diff
@@ -846,7 +885,28 @@ class SparkConnectPlanner(val session: SparkSession) {
-    logical.Filter(condition = transformExpression(rel.getCondition), child = baseRel)
+    val cond = rel.getCondition
+    cond.getExprTypeCase match {
+      case proto.Expression.ExprTypeCase.COMMON_INLINE_USER_DEFINED_FUNCTION
```
Looking at the proto, the minimal chance needed is:
1. Always pass the Scala UDF inputs and return types and 
2. Leave the CommonInlineUDF arguments as provided when available or not provided e.g. for typed filter function and map partitions function.

I feel this approach a. dose not make any proto change. b. does not add any extra data. c. can tell the typed filter udf clearly from others.
`col("*")` would do a similar result, but I would prefer avoid a magical marker if possible.

zhenlineo replied:
Changed to Col(*)

zhenlineo commented on connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:
```diff
@@ -482,27 +482,66 @@ class SparkConnectPlanner(val session: SparkSession) {
+      fun: proto.CommonInlineUserDefinedFunction,
+      child: LogicalPlan): LogicalPlan = {
+    val udf = fun.getScalarScalaUdf
+    val udfPacket =
```
Now that I have the input and return types, can I use reflection to find the encoders for them rather than use the encoders obtained from udfPacket?

I cannot easily test out using a `Person` class currently as I am missing the class sync logic. 

But in theory, if the class is synced to the server correctly, I should be able to directly use reflection to find the Person encoders, right?

hvanhovell replied:
No you cannot. That would mean you cannot use case classes or java beans. Rows would work though.

The reason we propagate types is to make the sure the engine can supply correctly typed data.

zhenlineo committed:
Use Col(*) for filter func

zhenlineo commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/UserDefinedFunctionE2ETestSuite.scala:
```diff
@@ -0,0 +1,131 @@
+    assert(longs == Arrays.asList[Long](0, 1, 2))
+  }
+
+  test("filter with col(*)") {
```
Note this test will run the udf as typed filter instead.

hvanhovell commented on connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -2468,6 +2470,110 @@ class Dataset[T] private[sql] (
+    sparkSession.newDataset(outputEncoder) { builder =>
+      builder.getMapPartitionsBuilder
+        .setInput(plan.getRoot)
+        .setFunc(udf.apply().expr.getCommonInlineUserDefinedFunction)
```
Nit we should also `col("*")` here to be consistent. I know that on the server side we ignore this.

hvanhovell reviewed (approved):
LGTM

hvanhovell commented:
Merging this one.
